packages/signature-controller/src/SignatureController.ts

```diff
@@ -0,0 +1,680 @@
+import EventEmitter from 'events';
+import {
+  MessageManager,
+  MessageParams,
+  MessageParamsMetamask,
+  PersonalMessageManager,
+  PersonalMessageParams,
+  PersonalMessageParamsMetamask,
+  TypedMessageManager,
+  TypedMessageParams,
+  TypedMessageParamsMetamask,
+  AbstractMessageManager,
+  AbstractMessage,
+  MessageManagerState,
+  AbstractMessageParams,
+  AbstractMessageParamsMetamask,
+  OriginalRequest,
+} from '@metamask/message-manager';
+import { ethErrors } from 'eth-rpc-errors';
+import { bufferToHex } from 'ethereumjs-util';
+
+import {
+  BaseControllerV2,
+  RestrictedControllerMessenger,
+} from '@metamask/base-controller';
+import { Patch } from 'immer';
+import {
+  AcceptRequest,
+  AddApprovalRequest,
+  RejectRequest,
+} from '@metamask/approval-controller';
+import { ORIGIN_METAMASK } from '@metamask/controller-utils';
+
+const controllerName = 'SignatureController';
+const methodNameSign = 'eth_sign';
+const methodNamePersonalSign = 'personal_sign';
+const methodNameTypedSign = 'eth_signTypedData';
+
+const stateMetadata = {
+  unapprovedMsgs: { persist: false, anonymous: false },
+  unapprovedPersonalMsgs: { persist: false, anonymous: false },
+  unapprovedTypedMessages: { persist: false, anonymous: false },
+  unapprovedMsgCount: { persist: false, anonymous: false },
+  unapprovedPersonalMsgCount: { persist: false, anonymous: false },
+  unapprovedTypedMessagesCount: { persist: false, anonymous: false },
+};
+
+const getDefaultState = () => ({
+  unapprovedMsgs: {},
+  unapprovedPersonalMsgs: {},
+  unapprovedTypedMessages: {},
+  unapprovedMsgCount: 0,
+  unapprovedPersonalMsgCount: 0,
+  unapprovedTypedMessagesCount: 0,
+});
+
+type CoreMessage = AbstractMessage & {
+  messageParams: AbstractMessageParams;
+};
+
+type StateMessage = Required<AbstractMessage> & {
+  msgParams: Required<AbstractMessageParams>;
+  securityProviderResponse: any;
```

Comment:
Didn't you need to Omit `securityProviderResponse` from AbstractMessage? That would save me a lot of time.

Reply:
Now that I think about it, the `Omit` suggestion was good for the encryption public key message because there isn't one, but in this case, that response is sent to the UI, so it should be there.

When we do the ticket about improving MessageManager types we can tackle all these details better.

Reply:
Forgot about this, I've removed it completely for clarity and fixed the type in `AbstractMessageManager` to remove the type error.

I've also removed the use of the draft state which was ultimately causing the infinite error.